### ISSUE TITLE
Implement projection and filter pushdown to Parquet

### ISSUE THREAD

**Issue by sfkeller** [OSGeo/gdal#11418]
### Feature description

This concept is a read-only feature well known in databases, big data and distributed computing. Filtering and projection pushdown provide significant performance benefits if the Parquet file is prepared accordingly. 

Projection pushdown means that when you query a Parquet file, the driver only reads the columns required for the query. 

Filter pushdown means that when you apply a filter to a column scanned from a Parquet file, the filter is pushed down and can be used to skip parts of the file using the built-in zone maps of the Parquet file. Note that this depends on whether the Parquet file contains such zone maps.

### Additional context

See for example https://duckdb.org/docs/data/parquet/overview#partial-reading. 

I'm not sure if it's that "filter pushdown" what @rouault meant when he wrote in #10887 "... I believe especially if the result set is small compared to the table size, despite my attempts at translating OGR SQL to Arrow filtering capabilities".

I may be able to contribute - but I'd like to know if this is feasible?

**Comment by sfkeller** [comment]
Thanks for the quick answer. Just to understand you correctly: The parquet reader already does filtering and projection pushdown? (sorry my ignorance about the implementation where I don't yet see what the typical dependency to Arrow as a row format means).

E.g. like with following test query from https://docs.overturemaps.org/getting-data/duckdb/ not yet adapted to OGR, where the attributes id, name and geometry are "projected" and the WHERE statement "filters" optimized - including hive_partitioning and zone maps):

```
SELECT
  id,
  names.primary as name,
  geometry                            -- the GEOMETRY type to be implemted 
FROM read_parquet('s3://overturemaps-us-west-2/release/2024-10-23.0/theme=places/type=place/*', filename=true, hive_partitioning=1)
WHERE 
  bbox.xmin BETWEEN -75 AND -73       -- Only use the bbox miny/miny values 
  AND bbox.ymin BETWEEN 40 AND 41     -- because they are point geometries.
  AND categories.primary = 'pizza_restaurant':
```

**Comment by rouault** [comment]
> The parquet reader already does filtering and projection pushdown?

yes, for partitionned datasets like OvertureMaps, the "PARQUET:" dataset mode (https://gdal.org/en/stable/drivers/vector/parquet.html#dataset-partitioning-read-support) is necessary used and attribute and spatial filter pushdown is implemented using https://arrow.apache.org/docs/cpp/dataset.html#filtering-data . I believe the reason for DuckDB being much faster than Arrow is a smarter implementation. I don't think there's anything that can be done on GDAL side (I may have missed something of course)

**Comment by rouault** [comment]
https://github.com/OSGeo/gdal/pull/11459 implements attribute and spatial filter pushdown for Parquet through the new to 3.11dev ADBC driver

**Comment by dwilson1988** [comment]
@rouault - I may be missing something, but whenever I've tried to create a geoparquet file using GDAL, they are missing Column and Offset indexes - is this expected? I don't see how efficient filter pushdown would be possible without these. I, of course, could be mistaken here.

**Comment by rouault** [comment]
> they are missing Column and Offset indexes

what do you mean exactly by Column and Offset indexes ?

**Comment by dwilson1988** [comment]
The page indexes for each column chunk: https://parquet.apache.org/docs/file-format/pageindex/

**Comment by rouault** [comment]
> The page indexes for each column chunk: https://parquet.apache.org/docs/file-format/pageindex/

thanks.  The capability was added in libarrow 12 per https://github.com/apache/arrow/pull/34054 , but required to be explicitly enabled. Will be done per https://github.com/OSGeo/gdal/pull/11565

**Comment by dwilson1988** [comment]
Fantastic! Thanks for the speedy response!

**Comment by sfkeller** [comment]
Hi Even @rouault : Any update in this issue? (Sorry for asking, I'm "just" curious since I'm doing stuff with OGR and GeoParquet with OSM data and would like to be up-to-date).

**Comment by rouault** [comment]
> Any update in this issue?

I'm not sure which update you expect. For me everything we can do on GDAL side (or at least that I'm aware of given libarrow/libparquet capabilities) is implemented. I believe this issue could be closed

**Comment by sfkeller** [comment]
Ok. Many thanks! So let's close it.
